### DISCUSSION TITLE
Partition of nvme as metadata device  of hdd

### DISCUSSION THREAD

**Discussion by zhucan** [rook/rook#13191] [Ideas]
At present, only can use a whole piece nvme or hdd device as metadata device of hdd to create osd， the nvme disk evenly partitioned by rook;  but we want  to use different partition of multi nvme device  as metadata device of hdd;

1. First partition all Nvme disks
2. Let the metadata device of the hdd disk specify the partition

My means. We can specify metadata of hdd, not evenly partitioned by rook.

**Comment by travisn** [comment]
To use partitions, you could follow the [cluster-on-pvc example](https://rook.github.io/docs/rook/v1.12/CRDs/Cluster/pvc-cluster/#dedicated-metadata-and-wal-device-for-osd-on-pvc). As long as you define the local PVs with the partitions, this should work.

**Reply by zhucan** [reply]
but we  don't want to use local  pvs, is there way?

**Reply by thenamehasbeentake** [reply]
Maybe it will be complicated to recovery when the OSD disk is damaged if using local pvs.

**Reply by travisn** [reply]
@zhucan If not using PVCs, it is only possible to use a single metadata device for all the OSDs on a node.

**Reply by zhucan** [reply]
is it possible to support  it? @travisn

**Reply by travisn** [reply]
@zhucan What exactly do you want to support? You want to specify a metadata partition for a single OSD instead of the whole disk? That would first require a change to `ceph-volume`, you would need to open a ceph tracker for that.

**Reply by zhucan** [reply]
@travisn  yes, we want to specify a metadata partition for a single OSD instead of the whole disk. I think "ceph-volume" has support it.

**Reply by travisn** [reply]
@zhucan If you can find the ceph-volume command for using the existing metadata partition, it would be helpful to consider supporting this in rook. I know others are interested in the feature too.

**Comment by zhucan** [comment]
and other question is: the nvme name is different from different node.

**Reply by travisn** [reply]
The `metadataDevice` name can be different if defining the `nodes` in the Cluster CR.

**Comment by microyahoo** [comment]
Rook currently only supports one metadata device per node, and then it is divided into multiple partitions, one for each osd. If one of the osd disks is broken, then I replace the disk with good one and add it back to the cluster, or remove it from the cluster and then add one good disk back again, what state will the cluster be in at this time? Can this osd automatically use the previous metadata device partition? Thanks. @travisn

**Reply by travisn** [reply]
No, the osd cannot currently use the previous metadata device partition. To replace a single OSD, currently all OSDs on the node with the metadata device would need to be wiped and replaced, unfortunately.

**Reply by zhucan** [reply]
@travisn all of osds needed rebuild?

**Reply by microyahoo** [reply]
I think the solution of wiping all osds on the node is unfeasible in a production environment. It will have a great impact on the cluster, which is equivalent to the entire node being down.

**Reply by microyahoo** [reply]
Are there any plans to improve? @travisn @satoru-takeuchi

**Reply by satoru-takeuchi** [reply]
At least I have no plan.

**Reply by zhucan** [reply]
1. Partition the disk first
2. In yaml, specify partitions for each data disk
3. create osd

I think it is more in line with actual usage scenarios and more convenient for operation and maintenance. @travisn @satoru-takeuchi

**Reply by travisn** [reply]
Certainly it is more desirable, but it is just not implemented today for non-pvc clusters.

**Reply by travisn** [reply]
#13240 opened to discuss a possible fix for this
